drivers/dma/dma_cavs.c

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <kernel.h>
+#include <board.h>
+#include <device.h>
+#include <init.h>
+#include <dma.h>
+
+#include "dma_cavs.h"
+
+#define SYS_LOG_DOMAIN "dev/dma_cavs"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_DMA_LEVEL
+#include <logging/sys_log.h>
+
+#define BYTE				(1)
+#define WORD				(2)
+#define DWORD				(4)
+
+/* CFG_LO */
+#define DW_CFG_CLASS(x)			(x << 29)
+/* CFG_HI */
+#define DW_CFGH_SRC_PER(x)		((x & 0xf) | ((x & 0x30) << 24))
+#define DW_CFGH_DST_PER(x)		(((x & 0xf) << 4) | ((x & 0x30) << 26))
+
+/* default initial setup register values */
+#define DW_CFG_LOW_DEF			0x0
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_DATA(dev) ((struct dw_dma_dev_data *const)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((const struct dw_dma_dev_cfg *const)(dev)->config->config_info)
+
+/* number of tries to wait for reset */
+#define DW_DMA_CFG_TRIES	10000
+#define INT_MASK_ALL		0xFF00
+
+static ALWAYS_INLINE void dw_write(u32_t dma_base, u32_t reg, u32_t value)
+{
+	*((volatile u32_t*)(dma_base + reg)) = value;
+}
+
+static ALWAYS_INLINE u32_t dw_read(u32_t dma_base, u32_t reg)
+{
+	return *((volatile u32_t*)(dma_base + reg));
+}
+
+static void dw_dma_isr(void *arg)
```

Comment:
Is this port in particular capable of SMP? If so, we might want to copy the DMA callbacks to local variables before looping through blocks/tfrs.

Reply:
Currently this port doesn't have SMP but we may need to support it in future. DMA callbacks are already copied to local variables. Can you let me know what is not copied here?